# Data Preprocessing - Filling NaN values
## Filling NaN values is a crucial step in data precprocessing as it gives Machine Learning models data to work with 
## In this notebook, we use the records with complete data to help fill in the NaN values using Linear Interpolation

### We start by importing the necessary libraries and loading the dataset into the notebook

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
df = pd.read_csv("../data_folder/sub-division_rainfall_act_dep_1901-2015.csv",
                encoding = "unicode_escape")
df.head()

SUBDIVISION       YEAR                 Parameter    JAN  \
0  ANDAMAN & NICOBAR ISLANDS  1901-2015                      Mean   49.2   
1  ANDAMAN & NICOBAR ISLANDS  1901-2015        Standard deviation   71.3   
2  ANDAMAN & NICOBAR ISLANDS  1901-2015  Coefficient of variation  144.9   
3  ANDAMAN & NICOBAR ISLANDS       1901                    Actual   49.2   
4  ANDAMAN & NICOBAR ISLANDS       1901      Percentage departure   -0.1   

     FEB    MAR   APR    MAY    JUN    JUL    AUG    SEP    OCT    NOV    DEC  \
0   27.6   30.0  72.2  355.6  471.4  397.5  400.5  431.3  289.5  233.0  153.3   
1   38.8   43.6  66.8  151.2  147.0  151.9  142.6  146.7   99.4  119.4  129.5   
2  140.4  145.5  92.5   42.5   31.2   38.2   35.6   34.0   34.3   51.2   84.5   
3   87.1   29.2   2.3  528.8  517.5  365.1  481.1  332.6  388.5  558.2   33.6   
4  215.9   -2.6 -96.8   48.7    9.8   -8.2   20.1  -22.9   34.2  139.5  -78.1   

   ANNUAL     JF    MAM    JJAS    OND  
0  2911.0   76.8  457.8  1700.7  675.8  
1   395.9   81.1  176.8   286.3  199.0  
2    13.6  105.6   38.6    16.8   29.4  
3  3373.2  136.3  560.3  1696.3  980.3  
4    15.9   77.5   22.4    -0.3   45.1

In [3]:
df = df.drop(['ANNUAL', 'JF', 'MAM',
       'JJAS', 'OND'], axis=1)
df.head()

SUBDIVISION       YEAR                 Parameter    JAN  \
0  ANDAMAN & NICOBAR ISLANDS  1901-2015                      Mean   49.2   
1  ANDAMAN & NICOBAR ISLANDS  1901-2015        Standard deviation   71.3   
2  ANDAMAN & NICOBAR ISLANDS  1901-2015  Coefficient of variation  144.9   
3  ANDAMAN & NICOBAR ISLANDS       1901                    Actual   49.2   
4  ANDAMAN & NICOBAR ISLANDS       1901      Percentage departure   -0.1   

     FEB    MAR   APR    MAY    JUN    JUL    AUG    SEP    OCT    NOV    DEC  
0   27.6   30.0  72.2  355.6  471.4  397.5  400.5  431.3  289.5  233.0  153.3  
1   38.8   43.6  66.8  151.2  147.0  151.9  142.6  146.7   99.4  119.4  129.5  
2  140.4  145.5  92.5   42.5   31.2   38.2   35.6   34.0   34.3   51.2   84.5  
3   87.1   29.2   2.3  528.8  517.5  365.1  481.1  332.6  388.5  558.2   33.6  
4  215.9   -2.6 -96.8   48.7    9.8   -8.2   20.1  -22.9   34.2  139.5  -78.1

In [4]:
print(df.isnull().sum().sum())

297


We can see that the current dataset has 297 NaN values that we need to take care of

In [5]:
df_actual = pd.DataFrame(columns = ['SUBDIVISION', 'YEAR', 'Parameter', 'JAN', 'FEB', 'MAR', 'APR', 'MAY',
       'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC'])
df_per_dep = df_actual
df_no_dis = df_actual

In [6]:
for i in range(0, df.shape[0]):
    if df.loc[i]["Parameter"] == "Actual":
        df_actual = df_actual.append(df.loc[i])
    elif df.loc[i]["Parameter"] == "Percentage departure":
        df_per_dep = df_per_dep.append(df.loc[i])
    elif df.loc[i]["Parameter"] == "No. of districts":
        df_no_dis = df_no_dis.append(df.loc[i])

We split the data frame into 3 categories accroding to their paramter - <br>
1) Actual <br>
2) Percentage Departure <br>
3) No. of districts <br>

In [7]:
print(df_actual.shape)
print(df_per_dep.shape)
print(df_per_dep.shape)

(4116, 15)
(4116, 15)
(4116, 15)


In [8]:
print(df_actual.isnull().sum().sum())
print(df_per_dep.isnull().sum().sum())
print(df_no_dis.isnull().sum().sum())
print(df.isnull().sum().sum())

70
227
0
297


We can see the break up on NaN values in each category <br>
We use the interpolate function of pandas to fill in these NaN values with the average of the year before and year after. <br>
Eg - if 1923 has an NaN value, it will be replaced with the average of 1922 and 1924

In [9]:
df_actual = df_actual.interpolate(method='linear', limit_direction='both')
df_per_dep = df_per_dep.interpolate(method='linear', limit_direction='both')
df_no_dis = df_no_dis.interpolate(method='linear', limit_direction='both')

In [10]:
print(df_actual.isnull().sum().sum())
print(df_per_dep.isnull().sum().sum())
print(df_no_dis.isnull().sum().sum())

0
0
0


As we can see, there are now no NaN values!

In [11]:
df_actual.to_csv("../data_folder/rainfall_actual.csv")
df_per_dep.to_csv("../data_folder/rainfall_per_dep.csv")
df_no_dis.to_csv("../data_folder/rainfall_no_dis.csv")

We finish by saving these dataframes as csv workbooks so we can use them for further analysis

In [12]:
df_actual

SUBDIVISION  YEAR Parameter    JAN    FEB    MAR    APR  \
3      ANDAMAN & NICOBAR ISLANDS  1901    Actual   49.2   87.1   29.2    2.3   
6      ANDAMAN & NICOBAR ISLANDS  1902    Actual    0.0  159.8   12.2    0.0   
9      ANDAMAN & NICOBAR ISLANDS  1903    Actual   12.7  144.0    0.0    1.0   
12     ANDAMAN & NICOBAR ISLANDS  1904    Actual    9.4   14.7    0.0  202.4   
15     ANDAMAN & NICOBAR ISLANDS  1905    Actual    1.3    0.0    3.3   26.9   
18     ANDAMAN & NICOBAR ISLANDS  1906    Actual   36.6    0.0    0.0    0.0   
21     ANDAMAN & NICOBAR ISLANDS  1907    Actual  110.7    0.0  113.3   21.6   
24     ANDAMAN & NICOBAR ISLANDS  1908    Actual   20.9   85.1    0.0   29.0   
27     ANDAMAN & NICOBAR ISLANDS  1910    Actual   26.6   22.7  206.3   89.3   
30     ANDAMAN & NICOBAR ISLANDS  1911    Actual    0.0    8.4    0.0  122.5   
33     ANDAMAN & NICOBAR ISLANDS  1912    Actual  583.7    0.8    0.0   21.9   
36     ANDAMAN & NICOBAR ISLANDS  1913    Actual   84.8    0.5    1.3    2.5   
39     ANDAMAN & NICOBAR ISLANDS  1914    Actual    0.0    0.0    0.0   37.7   
42     ANDAMAN & NICOBAR ISLANDS  1915    Actual   45.0   56.7   33.3   40.9   
45     ANDAMAN & NICOBAR ISLANDS  1916    Actual    0.0    0.0    0.0    0.5   
48     ANDAMAN & NICOBAR ISLANDS  1917    Actual    8.0    3.6  112.0    4.5   
51     ANDAMAN & NICOBAR ISLANDS  1918    Actual   77.4    6.9   11.4   10.7   
54     ANDAMAN & NICOBAR ISLANDS  1919    Actual   10.2   18.0    0.0   35.5   
57     ANDAMAN & NICOBAR ISLANDS  1920    Actual  122.3    7.4    3.1   13.0   
60     ANDAMAN & NICOBAR ISLANDS  1921    Actual   13.2    3.1    0.0   37.5   
63     ANDAMAN & NICOBAR ISLANDS  1922    Actual  245.3   34.3   15.6  323.1   
66     ANDAMAN & NICOBAR ISLANDS  1923    Actual   79.5    0.0   15.2   91.3   
69     ANDAMAN & NICOBAR ISLANDS  1924    Actual   28.7    0.0   14.8   89.7   
72     ANDAMAN & NICOBAR ISLANDS  1925    Actual   36.6    0.0    8.6   50.4   
75     ANDAMAN & NICOBAR ISLANDS  1926    Actual  122.1    0.0    0.0    0.5   
78     ANDAMAN & NICOBAR ISLANDS  1927    Actual    3.0   17.5   17.8  108.6   
81     ANDAMAN & NICOBAR ISLANDS  1928    Actual   50.9   67.6   80.7  129.3   
84     ANDAMAN & NICOBAR ISLANDS  1929    Actual   74.2  118.4  129.2   69.8   
87     ANDAMAN & NICOBAR ISLANDS  1930    Actual   87.4  105.4  131.2   10.9   
90     ANDAMAN & NICOBAR ISLANDS  1931    Actual   25.3    0.0    2.5    2.5   
...                          ...   ...       ...    ...    ...    ...    ...   
12366         WEST UTTAR PRADESH  1986    Actual   14.8   46.5    5.7    3.9   
12369         WEST UTTAR PRADESH  1987    Actual   15.2   13.1    2.7    1.8   
12372         WEST UTTAR PRADESH  1988    Actual    3.5   11.4   26.6    8.0   
12375         WEST UTTAR PRADESH  1989    Actual   24.6    1.1   15.4    0.0   
12378         WEST UTTAR PRADESH  1990    Actual    0.0   67.5    9.2    0.7   
12381         WEST UTTAR PRADESH  1991    Actual    2.9   18.9    7.5    2.9   
12384         WEST UTTAR PRADESH  1992    Actual   14.2    9.7    4.0    0.3   
12387         WEST UTTAR PRADESH  1993    Actual    3.7   13.5   21.2    1.0   
12390         WEST UTTAR PRADESH  1994    Actual   24.0   17.3    0.0   12.4   
12393         WEST UTTAR PRADESH  1995    Actual   30.1   11.2    7.9    3.7   
12396         WEST UTTAR PRADESH  1996    Actual   24.8   55.8    2.7    0.4   
12399         WEST UTTAR PRADESH  1997    Actual    3.8    0.7    2.4   19.2   
12402         WEST UTTAR PRADESH  1998    Actual    3.3    8.1   23.5    9.0   
12405         WEST UTTAR PRADESH  1999    Actual   19.3    1.1    0.0    0.0   
12408         WEST UTTAR PRADESH  2000    Actual    8.8   22.9    1.7    2.1   
12411         WEST UTTAR PRADESH  2001    Actual    2.5    2.3    2.6   12.0   
12414         WEST UTTAR PRADESH  2002    Actual   16.8   37.6    2.0    0.7   
12417         WEST UTTAR PRADESH  2003    Actual   18.6   40.4    2.5    4.4   
12420         WEST UT